lib/functions/imtd-process.js

```diff
@@ -0,0 +1,116 @@
+const parseThresholds = require('../models/parse-thresholds')
+const axios = require('axios')
+const logger = require('../helpers/logging')
+const pg = require('../helpers/db')
+
+// Inspection suggests max value is 9999, increase by order of magnitude for safety
+const MAX_RLOI_ID = 99999
+
+async function deleteThresholds (stationId) {
+  try {
+    await pg('station_imtd_threshold').where({ station_id: stationId }).delete()
+    logger.info(`Deleted thresholds for RLOI id ${stationId}`)
+  } catch (error) {
+    logger.error(`Error deleting thresholds for station ${stationId}`, error)
+    throw error
+  }
+}
+
+async function insertThresholds (stationId, thresholds) {
+  try {
+    const mappedThresholds = thresholds.map(t => {
+      return {
+        station_id: stationId,
+        fwis_code: t.floodWarningArea,
+        fwis_type: t.floodWarningType,
+        direction: t.direction,
+        value: t.level
+      }
+    })
+    await pg.transaction(async trx => {
+      await trx('station_imtd_threshold').where({ station_id: stationId }).delete()
+      await trx('station_imtd_threshold').insert(mappedThresholds)
+      logger.info(`Processed ${mappedThresholds.length} thresholds for RLOI id ${stationId}`)
+    })
+  } catch (error) {
+    logger.error(`Database error processing thresholds for station ${stationId}`, error)
+    throw error
+  }
+}
+
+async function getImtdApiResponse (stationId) {
+  const hostname = 'imfs-prd1-thresholds-api.azurewebsites.net'
+  try {
+    return await axios.get(`https://${hostname}/Location/${stationId}?version=2`)
+  } catch (error) {
+    if (error.response?.status === 404) {
+      logger.info(`Station ${stationId} not found (HTTP Status: 404)`)
+    } else {
+      const message = error.response?.status ? `HTTP Status: ${error.response.status}` : `Error: ${error.message}`
+      throw Error(`IMTD API request for station ${stationId} failed (${message})`)
+    }
+    return {}
+  }
+}
+
+async function getIMTDThresholds (stationId) {
+  const response = await getImtdApiResponse(stationId)
+  if (response.data) {
+    return parseThresholds(response.data[0].TimeSeriesMetaData)
+  }
+  return []
+}
+
+async function getData (stationId) {
+  try {
+    const thresholds = await getIMTDThresholds(stationId)
+    if (thresholds.length > 0) {
+      await insertThresholds(stationId, thresholds)
+    } else {
+      await deleteThresholds(stationId)
+    }
+  } catch (error) {
+    logger.error(`Could not process data for station ${stationId} (${error.message})`)
+  }
+}
+
+async function getRloiIds () {
+  // NOTE: ability to support a subset of all id's is for use in testing and also to allow us to run multiple
```

Comment:
A auto scalable alternative would be to use properties in the lambda's invocation event to use sql's LIMIT and OFFSET. 

This would work by adding an `offset` property in the invocation `event` with a default value of 0 and a limit set in an env var, set to something like 1000.
The scheduled invocation has no offset property to gets the first 1000 rloi_ids, when it finishes it calls `invokeFunction` (via the aws sdk) on itself, this time with an offset of 1000 (by adding the offset it was provided (0) to the limit (1000). The functions keep doing this until the call to the database returns no results.

Reply:
I've removed the ability to specify a subset of RLOI id's. There is no hard requirement for it at the moment and it wasn't used during testing.

We can revisit it if run time duration becomes an issue in the future.